server/src/main/java/org/elasticsearch/action/bulk/IncrementalBulkService.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.bulk;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.support.ActiveShardCount;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.core.TimeValue;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class IncrementalBulkService {
+
+    private final Client client;
+
+    public IncrementalBulkService(Client client) {
+        this.client = client;
+    }
+
+    public Handler newBulkRequest() {
+        return newBulkRequest(null, null, null);
+    }
+
+    public Handler newBulkRequest(@Nullable String waitForActiveShards, @Nullable TimeValue timeout, @Nullable String refresh) {
+        return new Handler(client, waitForActiveShards, timeout, refresh);
+    }
+
+    public static class Handler {
+
+        private final Client client;
+        private final ActiveShardCount waitForActiveShards;
+        private final TimeValue timeout;
+        private final String refresh;
+
+        private final ArrayList<Releasable> releasables = new ArrayList<>(4);
+        private final ArrayList<BulkResponse> responses = new ArrayList<>(2);
+        private Exception topLevelFailure = null;
+        private BulkRequest bulkRequest = null;
+
+        private Handler(Client client, @Nullable String waitForActiveShards, @Nullable TimeValue timeout, @Nullable String refresh) {
+            this.client = client;
+            this.waitForActiveShards = waitForActiveShards != null ? ActiveShardCount.parseString(waitForActiveShards) : null;
+            this.timeout = timeout;
+            this.refresh = refresh;
+            createNewBulkRequest();
+        }
+
+        public void addItems(List<DocWriteRequest<?>> items, Releasable releasable, Runnable nextItems) {
+            if (topLevelFailure != null) {
+                assert releasables.isEmpty();
+                assert bulkRequest == null;
+                Releasables.close(releasable);
+                nextItems.run();
+            } else {
+                internalAddItems(items, releasable);
+
+                if (shouldBackOff()) {
```

Comment:
I dont think shouldBackoff need to be part of PR. Even as placeholder. I would imagine backoff would be some sort of future that will tell we no longer have memory pressure so we can continue consuming stream. In current case we keep reading stream on backoff. Something like
```java
memoryAwareObject.runWhenEnoughMemory(() -> {
 processItems();
 nextItems.run();
})
```

Reply:
> I dont think shouldBackoff need to be part of PR. Even as placeholder.

The current PR splits bulks at 16 items. That is not what we will do, but without any splitting there is no way to test or exercise the incremental logic which this PR is focused on implementing. The eventually splitting logic will (as you note) be something memory related.

> I would imagine backoff would be some sort of future that will tell we no longer have memory pressure so we can continue consuming stream.

Once we submit an incremental bulk request, we must wait for the response to ensure that we execute operations in order. At least for the first iterations on this.